app/code/Magento/Paypal/view/frontend/web/js/in-context/button.js

```diff
@@ -18,11 +18,18 @@ define([
         /** @inheritdoc */
         initialize: function (config, element) {
             var cart = customerData.get('cart'),
-                customer = customerData.get('customer');
+                customer = customerData.get('customer'),
+                updateDeclinePayment = function () {
+                    this.declinePayment = !customer().firstname && !cart().isGuestCheckoutAllowed;
+                }.bind(this);
 
             this._super();
             this.renderPayPalButtons(element);
-            this.declinePayment = !customer().firstname && !cart().isGuestCheckoutAllowed;
```

Comment:
Sometimes the `customer()` and `cart()` were empty objects because they weren't initialized yet. To fix this behavior, we're updating the value that we have right now and triggering re-calculation of this value on change to any of them.

Our local tests showed that this fixes the issue.